src/Winton.Extensions.Configuration.Consul/ConsulConfigurationProvider.cs

```diff
@@ -156,7 +156,7 @@ private async Task PollingLoop(CancellationToken cancellationToken)
 
         private void SetData(QueryResult<KVPair[]> result)
         {
-            Data = result.ToConfigDictionary(_source.ConvertConsulKVPairToConfig);
+            Data = result.HasValue() ? result.ToConfigDictionary(_source.ConvertConsulKVPairToConfig) : null;
```

Comment:
I think we need to include a check here and only set `Data` to `null` when `_source.IsOptional`. Otherwise we'll be unsetting required data which I think is different to the design of the other configuration providers.

It's a bit of a pain because we'll end up duplicating some of the checking that happens in the initial `DoLoad`. This isn't a big problem, but one potential solution would be to make this method return `bool` to indicate whether it was able to update the `Data` or not and then in `DoLoad` throw if setting fails (because the data was missing and required).

Reply:
I will try to do some refactoring to simplify the code.

Reply:
Sorry, I'm not sure if my first comment was clear as I think we're both in agreement that `DoLoad` is only called once when it first loads and then `PollingLoop` is responsible for monitoring for changes and reloading. 

What I was referring to is the fact that in `PollingLoop` we need to make sure that don't set `Data` to `null` if `result.HasValue()` is `false` and `_source.IsOptional` is `false`. So we need something similar to the code that's in `DoLoad` but it shouldn't throw in the case when the result doesn't have a value and it is required, it should just silently ignore this and leave the `Data` as it was.

Reply:
OK. I think the best approach (in terms of code size and complexity) would be to just inline the SetData method and add null logic only to the PollingLoop version.

Reply:
Made the commit